File: solutions/c/luhn/1/luhn.c
#include "luhn.h"

#include <ctype.h>

bool luhn(const char* num) {
    // find the last char
    const char* p = num;
    while (*p) ++p;
    --p;

    int checksum = 0;
    int digits = 0;
    for (; p >= num; --p) {
        if (isspace(*p)) continue;
        if (!isdigit(*p)) return false;
        digits += 1;
        char digit = *p - '0';
        if (digits % 2 == 0) {
            digit *= 2;
            if (digit > 9) digit -= 9;
        }
        checksum += digit;
    }
    return digits >= 2 && checksum % 10 == 0;
}

File: solutions/haskell/luhn/1/src/Luhn.hs
module Luhn (isValid) where

import Data.Char (digitToInt, isDigit)

isValid :: String -> Bool
isValid cardNum = length digits >= 2 && checksum `mod` 10 == 0
  where
    digits = map digitToInt $ filter isDigit cardNum
    checksum = sum $ zipWith step [1 :: Int ..] $ reverse digits
    step n digit
        | odd n = digit
        | digitX2 > 9 = digitX2 - 9
        | otherwise = digitX2
      where
        digitX2 = digit * 2
